test: Bring back broken out-of-office tests, prevent modal button clickable before flicker, and fix timezone issue

ImBIOS opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #14287
- Fixes [CAL-3337](https://linear.app/calcom/issue/CAL-3337/bring-back-broken-out-of-office-tests-as-per-14286)

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set? none
- What are the minimal test data to have? none
- What is expected (happy path) to have (input and output)? i don't have idea what is it
- Any other important info that could help to test that PR:

```
yarn e2e out-of-office.e2e.ts
```

![image](https://github.com/user-attachments/assets/c84c5edb-176a-4c54-82bf-bb9f494d533f)
![image](https://github.com/user-attachments/assets/5c86ad3a-f9b6-4b78-ba96-03e082dbd9fa)
![image](https://github.com/user-attachments/assets/ac0cd666-6884-441d-aa05-6df0fec974eb)
![image](https://github.com/user-attachments/assets/9beac628-e4d9-4ead-a507-f3bb5912bbd3)

## Message

Hey everyone! 👋

I'm working on addressing the issue [CAL-3337] to bring back the broken out-of-office tests. Here's the lowdown on what's happening and how we can fix it! 🚀

#### Cause:
- There's a full page flicker during the hydration period, which causes the Modal state to reset. 😵‍💫
- Our e2e test, Playwright, is opening the modal at superhuman speed, but while waiting for the dropdown data to load, the modal gets reset by the flicker. This results in an error when the element isn't found. 😓

#### Known:
- Why does the flicker happen? Still a mystery! 🔍 I suspect this: https://github.com/calcom/cal.com/blob/5b7f2a6ffcfcda3aa9411fe482fd58519a76ed07/packages/ui/components/meta/Meta.tsx#L52

#### Potential Solutions:
1. **Change the test**:
   - Wait for certain text (like the title) to load, indicating the hydration is done and the flicker has passed. ⏳
2. **Change the code**:
   - **Fix the flicker cause**: If we can identify and resolve the root cause of the flicker, the problem will be solved! 🛠️
   - **Change add button with loading skeleton until flicker passes**: This ensures the UX isn't affected by the modal closing unexpectedly and prevents Playwright from picking the modal before flickering. ✨

Let's tackle this issue and make the tests robust again! 💪💻

If you have any ideas or insights on why the flicker happens, feel free to share! Collaboration is key! 🤝🔧

Note: I'm not using this, because it will require the component to become async, which makes me want to write a new component inside this file for this button, but I'm not doing it, I don't know why:

```ts
  const data = await utils.viewer.outOfOfficeEntriesList.ensureData();
```

ImBIOS committed:
test: prevent modal button clickable before flicker

ImBIOS committed:
Merge branch 'main' into test/bring-back-out-of-office-e2e

ImBIOS committed:
chore: add note to migrate to appDir to solve full page flicker issue

anikdhabal commented on apps/web/pages/settings/my-account/out-of-office/index.tsx:
```diff
@@ -375,20 +375,30 @@ const OutOfOfficePage = () => {
-            data-testid="add_entry_ooo">
-            <Icon name="plus" size={16} /> {t("add")}
-          </Button>
+          isBeforeFlicker ? (
+            <SkeletonText className="h-8 w-20" />
+          ) : (
+            <Button
+              color="primary"
+              className="flex w-20 items-center justify-between px-4"
+              onClick={() => setOpenModal(true)}
+              data-testid="add_entry_ooo">
+              <Icon name="plus" size={16} /> {t("add")}
+            </Button>
+          )
```
Nah, this does not look good for rendering the button conditionally. Instead, you should disable the button

ImBIOS replied:
Because it's non-interactive due to waiting for several data to be loaded, isn't the loading state better than just the disabled state? LMKWYT.

anikdhabal commented on apps/web/pages/settings/my-account/out-of-office/index.tsx:
```diff
@@ -375,20 +375,30 @@ const OutOfOfficePage = () => {
 
   const [openModal, setOpenModal] = useState(false);
+
+  const { data } = trpc.viewer.outOfOfficeReasonList.useQuery();
```
Can we just extract the` isPending` status and use it instead of this: `const isBeforeFlicker = data?.length === undefined || data.length <= 0;`

Also, I found that this query is called from two different components. I think it would be better to pass the data and isPending status to the <OutOfOfficeEntriesList /> component instead of calling it two time.
Lmk what you think

ImBIOS replied:
It seems `isPending` is enough, thanks for noticing.

![image](https://github.com/user-attachments/assets/4e6bc8ca-51eb-4d7d-8795-a050a54f0eca)

I don't think passing tRPC x React-Query data to a component that uses the same endpoint is the wisest decision because React-Query (even the default Next.js default fetch did this too) will handle the de-duplication using its Context. When we passed the value from the parent, the React Query Context and deduplication became useless. As you can see when we installed `react-query-devtools` the endpoint was just called once and distributed to all components using it.

anikdhabal committed:
Merge branch 'main' into test/bring-back-out-of-office-e2e

anikdhabal commented on apps/web/pages/settings/my-account/out-of-office/index.tsx:
```diff
@@ -375,20 +375,30 @@ const OutOfOfficePage = () => {
   const [openModal, setOpenModal] = useState(false);
+
+  const { data } = trpc.viewer.outOfOfficeReasonList.useQuery();
+  // TODO: The best solution is we should find the cause of full page flicker
+  // Taling about hydration problem, migrating to appDir might solve the problem automatically
+  // which will reset the Modal state, which makes bad UX and fails (superhuman click speed) E2E tests
```
Don't add too many comments; only include them if there's a TODO

ImBIOS replied:
I'm on it :smile:

ImBIOS committed:
Merge branch 'main' into test/bring-back-out-of-office-e2e

ImBIOS commented on apps/web/pages/settings/my-account/out-of-office/index.tsx:
```diff
@@ -375,20 +375,30 @@ const OutOfOfficePage = () => {
+  // TODO: The best solution is we should find the cause of full page flicker
+  // Taling about hydration problem, migrating to appDir might solve the problem automatically
+  // which will reset the Modal state, which makes bad UX and fails (superhuman click speed) E2E tests
+  const isBeforeFlicker = data?.length === undefined || data.length <= 0;
```
A hydration error from Next.js popped up before this fix sometimes, preventing the playwright from clicking anything and waiting infinitely. I suspect this because it stated it caused a flicker, but further investigation is suggested.
https://github.com/calcom/cal.com/blob/5b7f2a6ffcfcda3aa9411fe482fd58519a76ed07/packages/ui/components/meta/Meta.tsx#L52

ImBIOS committed:
chore: Improve OutOfOfficePage loading experience

Refactor the OutOfOfficePage component to improve the loading experience by using the `isPending` property from the `trpc.viewer.outOfOfficeReasonList.useQuery()` hook. This workaorund eliminates the full page flicker issue and provides a better user experience. The "Add" button now shows a loading state when the query is pending.

anikdhabal committed:
Update index.tsx

anikdhabal reviewed (changes requested):
Hey @ImBIOS ooo e2e test is still failing, can you figure it out and make the final thing

ImBIOS commented:
found the culprit, it's another problem but I'm looking into it:

```
    throw new TRPCError({ code: "BAD_REQUEST", message: "start_date_must_be_in_the_future" });
 ```

ImBIOS committed:
chore: Update outOfOffice.handler.ts to use input.offset instead of calculating offset

ImBIOS commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -25,7 +25,7 @@ export const outOfOfficeCreate = async ({ ctx, input }: TBookingRedirect) => {
   const inputStartTime = dayjs(startDate).startOf("day");
   const inputEndTime = dayjs(endDate).endOf("day");
-  const offset = dayjs(inputStartTime).utcOffset();
+  const offset = input.offset;
```
Damn, e2e test is real. This LOC culprit makes me laugh hard :rofl:

ImBIOS replied:
Let me check it one more time to be sure.

ImBIOS replied:
@anikdhabal I confirmed this issue consists of a flicker issue and that time start timezone offset issue. So these 2 fixes are required to enable these tests.

anikdhabal replied:
yes, please

anikdhabal committed:
fix e2e

anikdhabal committed:
Merge branch 'main' into test/bring-back-out-of-office-e2e

anikdhabal committed:
Update

anikdhabal committed:
revert

anikdhabal commented on packages/trpc/server/routers/loggedInViewer/outOfOffice.handler.ts:
```diff
@@ -25,7 +25,7 @@ export const outOfOfficeCreate = async ({ ctx, input }: TBookingRedirect) => {
   const inputStartTime = dayjs(startDate).startOf("day");
   const inputEndTime = dayjs(endDate).endOf("day");
-  const offset = dayjs(inputStartTime).utcOffset();
+  const offset = input.offset;
```
Hey @ImBIOS Isn't both gives the same result? I think it would be better to use server tz here .

ImBIOS replied:
`inputStartTime` doesn't have timezone offset data in it -it's just date and time-, that might be the purpose of the engineer who created the OOO feature sent this offset data to the server, but somehow he doesn't consume it in the server. This offset data is required to make the server know what's the time in UTC+0 -the server timezone-.

The server timezone is UTC+0 if it's in the GitHub Actions by default, and the frontend should be UTC+1 if it follows the playwright config which is `Europe/London`. This makes the `inputStartTime` generated by `dayjs` always 1 day before (1 hour before today's start time) the server time.

CMIIW.

If you want, I am also able to write more edge cases for this OOO feature so we do not just test the functional, but also the edge cases, e. g. different timezones, so we can test this offset data, if you agree I'll work on adding more tests for it by God's will.

anikdhabal replied:
Hey, we don't want today to be set as the start time; rather, the start date must be in the future. If we use the client-side generated offset, it will allow me to set today as the start time. That's why we use the server tz here. Let's not change it. Let's find a different way to fix the test.

anikdhabal replied:
@ImBIOS look the above message

ImBIOS replied:
Don't worry, I've been dedicating myself to working on it, since morning. I found something new and interesting.

Is it expected to be able to set today as startTime in UTC- but it cannot be set today for the startTime if we in UTC+ ???

ImBIOS replied:
> Don't worry, I've been dedicating myself to working on it, since morning. I found something new and interesting.
> 
> Is it expected to be able to set today as startTime in UTC- (e.g. America/Phoenix) but it cannot be set today for the startTime if we are in UTC+ (e.g. Asia/Jakarta) ??? Is this inconsistent behavior expected?
> 
> I found this interesting behavior when I was trying to add several E2E edge cases on timezone-difference users. And it happens when I use this:
> 
> ```
>   const offset = dayjs(inputStartTime).utcOffset();
> ```
> 
> I'm not testing with `input.offset` yet because you suggested. I revert to using server tz, so I'm working on testing the behavior when using server tz, to make sure it's working as expected.

Something weird, now I can't reproduce this finding. The tests always fail from morning to afternoon, but since 15 minutes ago, these tests always succeeded no matter what timezone I tested it from.

If God wills, I'll keep update to test it. But for now, et's use the server tz, I'll push it now.

ImBIOS replied:
Sorry for the drama, I've figured out the root cause. On my way to pushing it...

anikdhabal replied:
All good. You're doing great work

ImBIOS replied:
wait.. I wannna add some bonus to prevent this kind of bug happen ever again.

ImBIOS replied:
I added tests against this code for all time zones to make sure no timezone facing this issue.

ImBIOS replied:
@anikdhabal This bug might be solved and the e2e tests might passed, but for your information... some timezones have a 1-day range (e.g. 3 - 4) as the default value in the modal date range and I noticed some timezones will have a 2-day range (e.g. 3 - 5) as the default value, because it's another issue which -in my opinion- should be a patch bug, I'm not addressing this issue as of right now.

ImBIOS committed:
Merge branch 'calcom:main' into test/bring-back-out-of-office-e2e

ImBIOS committed:
chore: Fix offset calculation for out of office bookings

ImBIOS committed:
chore: Update date range calculation for out-of-office modal

ImBIOS committed:
chore: Remove console.log statement in out-of-office modal

ImBIOS committed:
chore: Update TimeZoneEnum with additional time zones

This commit updates the TimeZoneEnum in the types.ts file to include additional time zones. The new time zones added are:
- ID: Asia/Jakarta (UTC+7)
- SAMOA: Etc/GMT+12 (UTC-12)
- AMERICAN_SAMOA: Pacific/Pago_Pago (UTC-11)
- HAWAII: Pacific/Honolulu (UTC-10)
- ALASKA: America/Anchorage (UTC-9)
- PACIFIC_USA: America/Los_Angeles (UTC-8)
- CENTRAL_USA: America/Chicago (UTC-6)
- EASTERN_USA: America/New_York (UTC-5)
- VENEZUELA: America/Caracas (UTC-4)
- ARGENTINA: America/Buenos_Aires (UTC-3)
- BRAZIL: America/Noronha (UTC-2)
- AZORES: Atlantic/Azores (UTC-1)
- FRANCE: Europe/Paris (UTC+1)
- GREECE: Europe/Athens (UTC+2)
- RUSSIA: Europe/Moscow (UTC+3)
- UAE: Asia/Dubai (UTC+4)
- PAKISTAN: Asia/Karachi (UTC+5)
- BANGLADESH: Asia/Dhaka (UTC+6)
- SINGAPORE: Asia/Singapore (UTC+8)
- JAPAN: Asia/Tokyo (UTC+9)
- AUSTRALIA: Australia/Sydney (UTC+10)
- SOLOMON_ISLANDS: Pacific/Guadalcanal (UTC+11)
- NEW_ZEALAND: Pacific/Auckland (UTC+12)

The purpose of this update is to provide a comprehensive list of time zones for users to choose from when setting their time zone in the out-of-office feature.

ImBIOS commented on apps/web/playwright/fixtures/types.ts:
```diff
@@ -1,4 +1,27 @@
 export enum TimeZoneEnum {
-  USA = "America/Phoenix",
-  UK = "Europe/London",
+  USA = "America/Phoenix", // UTC-7
+  UK = "Europe/London", // UTC
+  ID = "Asia/Jakarta", // UTC+7
+  SAMOA = "Etc/GMT+12", // UTC-12
+  AMERICAN_SAMOA = "Pacific/Pago_Pago", // UTC-11
+  HAWAII = "Pacific/Honolulu", // UTC-10
+  ALASKA = "America/Anchorage", // UTC-9
+  PACIFIC_USA = "America/Los_Angeles", // UTC-8
+  CENTRAL_USA = "America/Chicago", // UTC-6
+  EASTERN_USA = "America/New_York", // UTC-5
+  VENEZUELA = "America/Caracas", // UTC-4
+  ARGENTINA = "America/Buenos_Aires", // UTC-3
+  BRAZIL = "America/Noronha", // UTC-2
+  AZORES = "Atlantic/Azores", // UTC-1
+  FRANCE = "Europe/Paris", // UTC+1
+  GREECE = "Europe/Athens", // UTC+2
+  RUSSIA = "Europe/Moscow", // UTC+3
+  UAE = "Asia/Dubai", // UTC+4
+  PAKISTAN = "Asia/Karachi", // UTC+5
+  BANGLADESH = "Asia/Dhaka", // UTC+6
+  SINGAPORE = "Asia/Singapore", // UTC+8
+  JAPAN = "Asia/Tokyo", // UTC+9
+  AUSTRALIA = "Australia/Sydney", // UTC+10
+  SOLOMON_ISLANDS = "Pacific/Guadalcanal", // UTC+11
+  NEW_ZEALAND = "Pacific/Auckland", // UTC+12
 }
```
@anikdhabal I'm just following the existing naming format for the enums, but I recommend using timezone explicit as the key e.g. `UTC`, `UTC_PLUS_7`, etc. for better readability. Or maybe for this test, we can create new enums (?)

ImBIOS committed:
Merge branch 'main' into test/bring-back-out-of-office-e2e

ImBIOS commented:
@anikdhabal any update on review?

ImBIOS committed:
Merge branch 'main' into test/bring-back-out-of-office-e2e

anikdhabal committed:
update

anikdhabal committed:
Merge branch 'main' into test/bring-back-out-of-office-e2e

anikdhabal committed:
fix

anikdhabal committed:
remove comment

anikdhabal reviewed (approved):
LGTM!!
